lib/network/level-net/sio-stream.js

```diff
@@ -0,0 +1,372 @@
+                return;
+            }
+            stream._onend();
+            cb(null);
+        });
+
+        this.socket.on('stream-error', payload => {
+            const { streamId, error } = payload;
+            debug('received stream-error event streamId', streamId,
+                 'error', error);
+            const stream = this.remoteStreams[streamId];
+            if (!stream) {
+                debug('no remote stream registered with id', streamId);
+                return;
+            }
+            stream._onerror(error);
+            delete this.remoteStreams[streamId];
+        });
+
+        this.socket.on('stream-hangup', payload => {
+            const { streamId } = payload;
+            debug('received stream-hangup event streamid', streamId);
+            const stream = this.localStreams[streamId];
+            if (!stream) {
+                debug('no local stream registered with id', streamId,
+                      '(may have already reached the end)');
+                return;
+            }
+            this.destroyStream(streamId);
+        });
+    }
+
+    encodeStreams(arg) {
+        if (!arg) {
+            return arg;
+        }
+        const isReadStream = (typeof(arg.pipe) === 'function'
+                              && typeof (arg.read) === 'function');
+        const isWriteStream = (typeof(arg.write) === 'function');
+
+        if (isReadStream || isWriteStream) {
+            const streamId = uuid();
+            const encodedStream = {
+                $streamId: streamId,
+                readable: isReadStream,
+                writable: isWriteStream,
+            };
+            debug('encoded stream:', encodedStream);
+            let transportStream;
+            if (isReadStream && isWriteStream) {
+                transportStream = new SIODuplexStream(this, streamId,
+                                                      this.maxPendingAck,
+                                                      this.ackTimeoutMs);
+            } else if (isReadStream) {
+                transportStream = new SIOOutputStream(this, streamId,
+                                                      this.maxPendingAck,
+                                                      this.ackTimeoutMs);
+            } else {
+                transportStream = new SIOInputStream(this, streamId);
+            }
+            if (isReadStream) {
+                arg.pipe(transportStream);
+                this.localStreams[streamId] = arg;
+                arg.on('end', () => {
+                    debug('stream end, removing from local streams',
+                          streamId);
+                    delete this.localStreams[streamId];
+                });
+            }
+            if (isWriteStream) {
+                transportStream.pipe(arg);
+            }
+            return encodedStream;
+        }
+        if (typeof(arg) === 'object') {
+            let encodedObj;
+            if (Array.isArray(arg)) {
```

Comment:
This is custom application parameters, so it may want to pass an array as a function's argument (e.g. the "future" db.batch() I think is using an array to pass the list of operations).